kernel/src/main/scala/cats/Semigroup.scala

```diff
@@ -0,0 +1,73 @@
+package cats
+
+import scala.{ specialized => sp }
+import scala.annotation.{ switch, tailrec }
+
+/**
+ * A semigroup is any set `A` with an associative operation (`combine`).
+ */
+trait Semigroup[@sp(Int, Long, Float, Double) A] extends Any with Serializable {
+
+  /**
+   * Associative operation taking which combines two values.
+   */
+  def combine(x: A, y: A): A
+
+  /**
+   * Return `a` combined with itself `n` times.
+   */
+  def combineN(a: A, n: Int): A =
+    if (n <= 0) throw new IllegalArgumentException("Repeated combining for semigroups must have n > 0")
```

Comment:
should we include `: found n = $n` here? I hate to get an exception like this that doesn't tell me what the actual value was.